File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/common/ShadowLord.java
package com.nyrds.pixeldungeon.mobs.common;

import com.nyrds.Packable;
import com.nyrds.pixeldungeon.ai.Fleeing;
import com.nyrds.pixeldungeon.ai.MobAi;
import com.nyrds.pixeldungeon.ai.Wandering;
import com.nyrds.pixeldungeon.levels.LevelTools;
import com.nyrds.pixeldungeon.levels.objects.LevelObjectsFactory;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.audio.Sample;
import com.nyrds.platform.util.StringsManager;
import com.watabou.noosa.tweeners.AlphaTweener;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.actors.Actor;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.blobs.Darkness;
import com.watabou.pixeldungeon.actors.blobs.Foliage;
import com.watabou.pixeldungeon.actors.mobs.Boss;
import com.watabou.pixeldungeon.actors.mobs.Mob;
import com.watabou.pixeldungeon.actors.mobs.Shadow;
import com.watabou.pixeldungeon.actors.mobs.WalkingType;
import com.watabou.pixeldungeon.actors.mobs.Wraith;
import com.watabou.pixeldungeon.effects.MagicMissile;
import com.watabou.pixeldungeon.effects.Speck;
import com.watabou.pixeldungeon.items.scrolls.ScrollOfWeaponUpgrade;
import com.watabou.pixeldungeon.items.wands.WandOfBlink;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.mechanics.Ballistica;
import com.watabou.pixeldungeon.scenes.GameScene;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.utils.Callback;

import org.jetbrains.annotations.NotNull;

/**
 * Created by DeadDie on 13.02.2016
 */
public class ShadowLord extends Boss implements IZapper {

	@Packable
	private boolean levelCreated         = false;

	@Packable
	private int cooldown                 = -1;

	public ShadowLord() {
		hp(ht(260));
		baseDefenseSkill = 40;
		baseAttackSkill  = 30;
		dmgMin = 30;
		dmgMax = 40;
		dr = 40;

		exp = 60;

		collect(new ScrollOfWeaponUpgrade());

		walkingType = WalkingType.ABSOLUTE;
	}

	private void spawnShadow() {
		int cell = level().getSolidCellNextTo(getPos());

		if (level().cellValid(cell)) {
			Mob mob = new Shadow();

			mob.setState(MobAi.getStateByClass(Wandering.class));

			WandOfBlink.appear(mob, cell);
		}
	}

	private void spawnWraith() {
		for (int i = 0; i < 4; i++) {
			int cell = level().getEmptyCellNextTo(getPos());

			if (level().cellValid(cell)) {
				Wraith.spawnAt(cell);
			}
		}
	}

	private void twistLevel() {

		if(!isAlive()) {
			return;
		}

		Level level = level();

		if(!levelCreated)
		{
			LevelTools.makeEmptyLevel(level, false);
			LevelTools.buildShadowLordMaze(level, 6);
			levelCreated = true;
		}


		int cell = level.getRandomLevelObjectPosition(LevelObjectsFactory.PEDESTAL);
		if (level.cellValid(cell)) {
			if (Actor.findChar(cell) == null) {
				Mob mob = Crystal.makeShadowLordCrystal();
				mob.setPos(cell);
				level.spawnMob(mob);

				final CharSprite sprite = mob.getSprite();

				sprite.alpha( 0 );
				GameScene.addToMobLayer(new AlphaTweener(sprite, 1, 0.4f));


				sprite.emitter().start( Speck.factory( Speck.LIGHT ), 0.2f, 3 );
				Sample.INSTANCE.play( Assets.SND_TELEPORT );

				int x, y;
				x = level.cellX(cell);
				y = level.cellY(cell);

				level.fillAreaWith(Darkness.class, x - 2, y - 2, 5, 5, 1);
			} else {
				damage(ht() / 9, this);
			}
		}
	}

	@Override
    public boolean canAttack(@NotNull Char enemy) {
		return level().distance(getPos(), enemy.getPos()) < 4 && Ballistica.cast(getPos(), enemy.getPos(), false, true) == enemy.getPos();
	}

	@Override
	protected void fx(int cell, Callback callback) {
		MagicMissile.purpleLight(getSprite().getParent(), getPos(), cell, callback);
		Sample.INSTANCE.play(Assets.SND_ZAP);
		getSprite().setVisible(false);
	}

	@Override
	public void damage(int dmg, @NotNull NamedEntityKind src) {
		super.damage(dmg, src);
		if (src != this) {
			if (dmg > 0 && cooldown < 0) {
				setState(MobAi.getStateByClass(Fleeing.class));

				Char jumpFrom = this;

				if (src instanceof Char) {
					jumpFrom = (Char)src;
				}
				Char finalJumpFrom = jumpFrom;

				CharUtils.blinkAway(this,
							(level, cell) -> level.distance(cell, finalJumpFrom.getPos())>3 && Actor.findChar(cell) == null);

				twistLevel();
				cooldown = 10;
			}
		}
	}

	@Override
    public boolean act() {
		cooldown--;

		if (getState() instanceof Fleeing) {
			if (cooldown < 0) {
				setState(MobAi.getStateByClass(Wandering.class));
				if (Math.random() < 0.7) {
					spawnWraith();
				} else {
					spawnShadow();
				}

                yell(StringsManager.getVar(R.string.ShadowLord_Intro));
			}
		}

		if (level().blobAmountAt(Darkness.class, getPos()) > 0 && hp() < ht()) {
			heal((ht() - hp()) / 4, level().blobs.get(Darkness.class));
		}

		if (level().blobAmountAt(Foliage.class, getPos()) > 0) {
			getSprite().emitter().burst(Speck.factory(Speck.BONE), 1);
			damage(1, this);
		}

		return super.act();
	}

	@Override
	public void die(@NotNull NamedEntityKind cause) {
		super.die(cause);
        yell(StringsManager.getVar(R.string.ShadowLord_Death));
		LevelTools.makeEmptyLevel(level(), false);
		level().unseal();
		Badges.validateBossSlain(Badges.Badge.SHADOW_LORD_SLAIN);
	}
}

File: scripts/mobs/ShadowLord.lua
--[[
  ShadowLord (batch 17d-5, was mobs/common/ShadowLord.java). Depth 25 boss,
  closes the boss series. Ranged-only attacker (3 cells + LOS; the zap
  visual rides the sprite zapEffect). Any damage makes it blink away in
  Fleeing state and twist the arena into a walled maze whose pedestals
  sprout kind-2 Crystals wrapped in Darkness - it heals standing in
  Darkness, chars in Foliage, and summons wraiths (or a Shadow) while
  fleeing. Death resets the arena. Carries no SkeletonKey, by the
  long-standing ShadowLord exclusion.
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"
local function carveAround(level, cell)
	if not level:cellValid(cell) then
		return
	end

	local x, y = level:cellX(cell), level:cellY(cell)
	for i = x - 1, x + 1 do
		for j = y - 1, y + 1 do
			if level:cellValid(i, j) and level:get(i, j) ~= RPD.Terrain.EMPTY then
				level:set(i, j, RPD.Terrain.EMPTY)
				RPD.GameScene:updateMap(level:cell(i, j))
			end
		end
	end
end

local function twist(self)
	if not self:isAlive() then
		return
	end

	local level = RPD.Dungeon.level
	local data = mob.restoreData(self)

	if not data.levelCreated then
		RPD.LevelTools:makeEmptyLevel(level, false)
		RPD.LevelTools:buildShadowLordMaze(level, 6)
		-- caveman: maze buries whoever stands on a grid line - hero sealed
		-- in solid terrain gets no line of fire, boss stalls forever (#23)
		carveAround(level, RPD.Dungeon.hero:getPos())
		carveAround(level, self:getPos())
		data.levelCreated = true
	end

	local cell = level:getRandomLevelObjectPosition("pedestal")
	if level:cellValid(cell) then
		if RPD.Actor:findChar(cell) == nil then
			local crystal = RPD.spawnMob("Crystal", cell, { var = 2 })
			local sprite = crystal:getSprite()
			sprite:alpha(0)
			-- java built the fade-in tweener directly; no lua ctor idiom yet
			RPD.GameScene:addToMobLayer(
				luajava.newInstance("com.watabou.noosa.tweeners.AlphaTweener", sprite, 1, 0.4))
			sprite:emitter():start(RPD.Sfx.Speck:factory(RPD.Sfx.Speck.LIGHT), 0.2, 3)
			RPD.playSound("snd_teleport")
			level:fillAreaWith(RPD.Blobs.Darkness, level:cellX(cell) - 2, level:cellY(cell) - 2, 5, 5, 1)
		else
			self:damage(math.floor(self:ht() / 9), self)
		end
	end
end

return mob.init{
	spawn = function(self, level)
		local bag = self:getBelongings()
		if not bag:getItem("ScrollOfWeaponUpgrade") then
			-- raw factory: java collected it without the treasury check
			self:collect(RPD.ItemFactory:itemByName("ScrollOfWeaponUpgrade"))
		end
	end,

	damage = function(self, dmg, src)
		local data = mob.restoreData(self)
		if src ~= self and dmg > 0 and (data.cooldownUntil or 0) <= self.time and dmg < self:hp() then
			data.cooldownUntil = self.time + 10

			local jumpFrom = self
			if RPD.CharUtils:isChar(src) then
				jumpFrom = src
			end
			RPD.CharUtils:blinkAwayFrom(self, jumpFrom, 3)

			twist(self)
		end
		-- no numeric return: the regular damage flow proceeds; the revenge AI
		-- re-asserts HUNTING over any state set here, so Fleeing rides act
	end,

	act = function(self)
		local data = mob.restoreData(self)
		local fleeing = self:getState():getTag() == "Fleeing"

		if data.cooldownUntil == nil then
			-- no active flee window
		elseif self.time < data.cooldownUntil then
			if not fleeing then
				RPD.setAi(self, "Fleeing")
			end
		else
			data.cooldownUntil = nil
			RPD.setAi(self, "Wandering")
			if math.random() < 0.7 then
				for _ = 1, 4 do
					local cell = RPD.Dungeon.level:getEmptyCellNextTo(self:getPos())
					if RPD.Dungeon.level:cellValid(cell) then
						RPD.CharUtils:spawnWraithAt(RPD.Dungeon.level, cell)
					end
				end
			else
				local cell = RPD.Dungeon.level:getSolidCellNextTo(self:getPos())
				if RPD.Dungeon.level:cellValid(cell) then
					local shadow = RPD.MobFactory:mobByName("Shadow")
					RPD.setAi(shadow, "Wandering")
					RPD.WandOfBlink:appear(shadow, cell)
				end
			end
			self:yell(RPD.textById("ShadowLord_Intro"))
		end

		local level = RPD.Dungeon.level
		if level:blobAmountAt(RPD.Blobs.Darkness, self:getPos()) > 0 and self:hp() < self:ht() then
			self:heal(math.floor((self:ht() - self:hp()) / 4), level:getBlobByName("Darkness"))
		end

		if level:blobAmountAt(RPD.Blobs.Foliage, self:getPos()) > 0 then
			self:getSprite():emitter():burst(RPD.Sfx.Speck:factory(RPD.Sfx.Speck.BONE), 1)
			self:damage(1, self)
		end
	end,

	die = function(self, cause)
		self:yell(RPD.textById("ShadowLord_Death"))
		RPD.LevelTools:makeEmptyLevel(RPD.Dungeon.level, false)
		RPD.CharUtils:validateBossSlain("SHADOW_LORD_SLAIN")
	end,
}
